sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,392 @@
+
+# Mocking Client Types
+
+Event Hubs is built to support unit testing with mocks, as described in the [Azure SDK Design Guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-mocking). This is an important feature of the library that allows developers to write tests that are completely focused on their own application logic, though they depend on the Event Hubs types.
+
+
+The examples that follow focus on scenarios likely to occur in applications, and demonstrate how to mock the Event Hubs types typically used in each scenario.
+
+## Publishing Events with the `EventHubProducerClient`
+
+When using batches to publish to Event Hubs, the key interactions with the `EventHubProducerClient` are calling `CreateBatchAsync` to create the batch and `SendAsync` to publish it.   Mocked batches accept a `List<EventData>` that is used as a backing store and can be inspected to verify that the application is adding events to the batch as expected.  The custom `TryAdd` callback can be used to control the decision for whether an event is accepted into the batch or is rejected.   
+
+This snippet demonstrates mocking the `EventHubProducerClient` using Moq, and creating `EventDataBatch` instances using the `EventHubsModelFactory`.
+
+```C# Snippet:EventHubs_Sample11_MockingEventDataBatch
+var mockProducer = new Mock<EventHubProducerClient>();
+
+// This sets the value returned by the EventDataBatch when accessing the Size property
+// It does not impact TryAdd on the mocked batch
+var batchSizeInBytes = 500;
+
+// Events added to the batch will be added here, but altering the events in this list will not change the
+// events in the batch, since they are stored inside the batch as well
+var backingList = new List<EventData>();
+
+// For illustrative purposes allow the batch to hold 3 events before
+// returning false.
+var batchCountThreshold = 3;
+
+var dataBatchMock = EventHubsModelFactory.EventDataBatch(
+        batchSizeBytes : batchSizeInBytes,
+        batchEventStore : backingList,
+        batchOptions : new CreateBatchOptions() { },
+        // The model factory allows the user to define a custom TryAdd callback, making
+        // it easy to test specific scenarios
+        eventData =>
+        {
+            var numElements = backingList.Count();
+            return numElements < batchCountThreshold;
+        });
+
+// Setting up a mock of the CreateBatchAsync method
+mockProducer.Setup(p => p.CreateBatchAsync(
+    It.IsAny<CancellationToken>()))
+    .ReturnsAsync(dataBatchMock);
```

Comment:
```suggestion
mockProducer
    .Setup(p => p.CreateBatchAsync(It.IsAny<CancellationToken>()))
    .ReturnsAsync(dataBatchMock);
```

Some folks struggle with the concept of method chaining.   If we keep the chained calls at the same indentation level, it can often help readability.